src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -39,19 +39,52 @@ public Datastore() {
 
   /** Stores the Message in Datastore. */
   public void storeMessage(Message message) {
+    // Do not post message if user is not message poster or not logged on
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      System.err.println("Invalid Credentials: attempt to post message while not logged in");
+      return;
+    }
+    String userEmail = userService.getCurrentUser().getEmail();
+    if (!message.getUser().equals(userEmail)) {
+      System.err.println(
+          "Invalid Credentials: User "
+              + userEmail
+              + " attempt to post message by "
+              + message.getUser());
+      return;
+    }
+
     Entity messageEntity = new Entity("Message", message.getId().toString());
     messageEntity.setProperty("user", message.getUser());
     messageEntity.setProperty("ID", message.getId().toString());
     messageEntity.setProperty("text", message.getText());
     messageEntity.setProperty("timestamp", message.getTimestamp());
     messageEntity.setProperty("country", message.getCountry());
     messageEntity.setProperty("category", message.getCategory());
-    messageEntity.setProperty("latitude", message.getLat());
-    messageEntity.setProperty("longitude", message.getLng());
+    messageEntity.setProperty("lat", message.getLat());
+    messageEntity.setProperty("lng", message.getLng());
```

Comment:
These were changed 1 - for consistancy and 2- because they were referenced this way below, which returned null as-it-was